noise/README.md

```diff
@@ -0,0 +1,400 @@
+are [defined][npf-message-format] with a maximum length of 65535 bytes.
+
+Noise messages are sent prefixed with their length in bytes, encoded as an
+unsigned 16-bit integer in network (big-endian) byte order.
+
+## Supported DH Functions, Ciphers, and Hash Functions
+
+The Noise framework allows protocol designers to choose from a small set of
+Diffie-Hellman functions, symmetric ciphers, and hash functions.
+
+For simplicity, noise-libp2p currently supports one function from each category:
+
+### Supported DH Functions
+
+noise-libp2p supports the 25519 DH functions [as defined in the Noise
+spec][npf-dh-25519].
+
+### Supported Cipher Functions
+
+noise-libp2p supports the ChaChaPoly cipher functions [as defined in the Noise
+spec][npf-cipher-chachapoly].
+
+### Supported Hash Functions
+
+noise-libp2p supports the SHA256 hash function [as defined in
+the Noise spec][npf-hash-sha256].
+
+## Supported Handshake Patterns
+
+Noise defines twelve [fundamental interactive handshake
+patterns][npf-fundamental-patterns] for exchanging public keys between parties and performing
+Diffie-Hellman computations.
+
+The patterns are named according to whether static keypairs are used, and if so,
+by what means each party gains knowledge of the other's static public key.
+
+From the [fundamental patterns section of the Noise spec][npf-fundamental-patterns]:
+
+    The first character refers to the initiator's static key:
+
+       * N = No static key for initiator
+       * K = Static key for initiator Known to responder
+       * X = Static key for initiator Xmitted ("transmitted") to responder
+       * I = Static key for initiator Immediately transmitted to responder, despite reduced or absent identity hiding
+
+    The second character refers to the responder's static key:
+
+      * N = No static key for responder
+      * K = Static key for responder Known to initiator
+      * X = Static key for responder Xmitted ("transmitted") to initiator
+
+noise-libp2p supports a [compound protocol][npf-compound-protocols] consisting
+of two fundamental handshake patterns, `XX` and `IK`. A variation of the `XX`
+pattern, `XXfallback` is used to gracefully recover from a failed `IK`
+handshake attempt.
+
+This pattern of using `XX`, `IK` and `XXfallback` is described in the Noise spec
+as the [Noise Pipes compound protocol][npf-noise-pipes].
+
+The`XX` pattern is used when the initiator (or in libp2p terms, the dialing peer)
+does not have knowledge of the static Noise key for the responder (the listening
+peer). As Noise static keys are distinct from libp2p identity keys, this is
+likely to be the case for the first Noise connection between two peers.
+
+Following a successful `XX` handshake, both peers may store the static Noise
+public key of the other. Future connections can attempt the `IK` handshake using
+the stored static key, which offers the benefits of zero-RTT encryption and
+fewer message round-trips.
```

Comment:
Am I right to assume that the security properties of zero-RTT data are equivalent to those of 0-RTT data in TLS1.3, i.e. can be replayed?

Reply:
@marten-seemann I think you're right that the IK handshake is susceptible to replay. It has this destination [security property](https://noiseprotocol.org/noise.html#payload-security-properties):

> Encryption to a known recipient, forward secrecy for sender compromise only, vulnerable to replay. This payload is encrypted based only on DHs involving the recipient's static key pair. If the recipient's static private key is compromised, even at a later date, this payload can be decrypted. This message can also be replayed, since there's no ephemeral contribution from the recipient.

Reply:
Thanks @yusefnapora. I think we should document these security properties carefully.
This is something that an implementation will need to account for in the API in one way or the other.

Reply:
Let’s be careful not to replicate the Noise spec here, but rather reference it.

Reply:
Note that lack of encryption of 0-RTT data or replay ability, in practice, do not pose a threat for us.

Remember this handshake takes place _in the context of a libp2p connection establishment_, 0-RTT data will presumably be used by multiselect 2.0 to preemptively negotiate the multiplexer. We should not expose 0-RTT facilities to the application itself, IMO.

Notes:

1. Replaying the handshake itself is fruitless. Even in the case of IX or IK (1-RTT, both susceptible to replay), because the attacker does not know the initiator's private keys, unless compromised (in which case, it's a different attack).
2. The 0-RTT data we'll presumably send has no side-effects, unlike in TLS 1.3 with HTTP, where you _could_ be sending an HTTP POST.